### PR TITLE
Custom instances of GenericPrincipal in WaffleAuthenticatorBase

### PR DESCRIPTION
**Pull request by Snap252**
refering:
https://github.com/Waffle/waffle/issues/561

my main target is it to get the roles not directly by the windows roles but by a further mechanism that gets some roles from a database query using the current windowsIdentity as a query parameter

### CODE REVIEW

File: Source/JNA/waffle-tomcat85/src/main/java/waffle/apache/NegotiateAuthenticator.java [hunk 1]
```diff
@@ -158,12 +159,11 @@ public boolean authenticate(final Request request, final HttpServletResponse res
             try {
                 this.log.debug("logged in user: {} ({})", windowsIdentity.getFqn(), windowsIdentity.getSidString());
 
-                final GenericWindowsPrincipal windowsPrincipal = new GenericWindowsPrincipal(windowsIdentity,
-                        this.principalFormat, this.roleFormat);
+                final GenericPrincipal genericPrincipal = createPrincipal(windowsIdentity);
 
-                this.log.debug("roles: {}", windowsPrincipal.getRolesString());
+                this.log.debug("roles: {}",  String.join(", ", genericPrincipal.getRoles()));
```
Comment (hazendaz) [inline Source/JNA/waffle-tomcat85/src/main/java/waffle/apache/NegotiateAuthenticator.java]: Please revert the change above.  getRolesString() already does the string joiner.

Comment (hazendaz) [reply Source/JNA/waffle-tomcat85/src/main/java/waffle/apache/NegotiateAuthenticator.java]: nevermind I see why now.

**Comment by hazendaz** [comment]
@Snap252 Can you also make the same change to all other negotiateAuthenticators?  That way it is supported throughout.  Thanks in advance.

**Comment by dblock** [comment]
This method should go into the parent class I believe and be invoked from https://github.com/Waffle/waffle/blob/master/Source/JNA/waffle-tomcat7/src/main/java/waffle/apache/WaffleAuthenticatorBase.java#L246 as well. It also needs tests.

File: Source/JNA/waffle-tomcat85/src/main/java/waffle/apache/NegotiateAuthenticator.java (hunk 1 above)
Comment (Snap252) [reply Source/JNA/waffle-tomcat85/src/main/java/waffle/apache/NegotiateAuthenticator.java]: Using the toString-Method log output would give the possibility to use more generic interfaces
this.log.debug("roles: {}",  String.join(", ", genericPrincipal.getRoles()));
->
this.log.debug("principal: {}",  principal.toString());

**Comment by Snap252** [comment]
Thx for the fast feedback.

- I pulled up the method to WaffleAuthenticatorBase
- I made the changes in all tomcat versions (but is there any difference between 8.5 and 9 - I can find only one deprecation in source code)

If you tell me that this is the right way I will write some testing code and enhance code coverage.

File: Source/JNA/waffle-tomcat8/src/main/java/waffle/apache/NegotiateAuthenticator.java [hunk 2]
```diff
@@ -158,12 +159,11 @@ public boolean authenticate(final Request request, final HttpServletResponse res
             try {
                 this.log.debug("logged in user: {} ({})", windowsIdentity.getFqn(), windowsIdentity.getSidString());
 
-                final GenericWindowsPrincipal windowsPrincipal = new GenericWindowsPrincipal(windowsIdentity,
-                        this.principalFormat, this.roleFormat);
+                final GenericPrincipal genericPrincipal = createPrincipal(windowsIdentity);
 
-                this.log.debug("roles: {}", windowsPrincipal.getRolesString());
+                this.log.debug("roles: {}",  String.join(", ", genericPrincipal.getRoles()));
```
Comment (dblock) [inline Source/JNA/waffle-tomcat8/src/main/java/waffle/apache/NegotiateAuthenticator.java]: Minor: one space too many.

Comment (hazendaz) [reply Source/JNA/waffle-tomcat8/src/main/java/waffle/apache/NegotiateAuthenticator.java]: Run a quick build, that will flush out that space or any others.

Comment (hazendaz) [reply Source/JNA/waffle-tomcat8/src/main/java/waffle/apache/NegotiateAuthenticator.java]: @Snap252 There isn't really much difference between the various integrations.  We probably could use an overall refactoring to make things better.  For now though, as new builds become available, because we brand the version, we supply a new copy.  So tomcat 9 is really just tomcat 8.5 with the minor variance you saw.  Spring 3, 4, and 5 copies I believe are completely identical.  Just haven't had time to really go in and look how to make it more streamlined.

**Comment by hazendaz** [comment]
@Snap252 Just for reference of when this might get released.  I'm waiting on spring security 5 to be officially released.  Once that is done, I plan to push out 1.9.0 version of waffle.  I'll include this work with it.  Thanks.

File: Source/JNA/waffle-tomcat8/src/main/java/waffle/apache/NegotiateAuthenticator.java (hunk 2 above)
Comment (Snap252) [reply Source/JNA/waffle-tomcat8/src/main/java/waffle/apache/NegotiateAuthenticator.java]: > Run a quick build, that will flush out that space or any others.
> - hazendaz

I already tried a maven test / maven install - but this kills me all the formatting (in existing and untouched files) - Am I doing something wrong, or is there anything I needed to know to keep formatting?

**Comment by hazendaz** [comment]
@Snap252 Can you add the same test case to tomcat7,8,9?  Spring has now released spring security 5 so now just wating on that piece from you to merge this then a few rounds of testing and this will be released.  Thanks.

**Comment by Snap252** [comment]
I was not sure if these tests are enough - and before copying it to the other subprojects I was waiting for your feedback :+1:  
But I think this is the result wanted. Let me know if anything is missing.
Thx

**Comment by hazendaz** [comment]
I think it's good enough for what I can see here.  I'll run some coverage after merging.  I'm merging now.  Thanks for getting this squared away.

**Comment by Snap252** [comment]
I feel a pretty guilty here,
> I think it's good enough for what I can see here.
I thought it is done. Can you tell me, what is still to do here?

Kind regards.

**Comment by hazendaz** [comment]
Just need to go through each tomcat and compare them to make sure all are in sync.  I've been short on time so haven't gotten back to this.  If you are able, great.  If not that is ok too but may be a few weeks before I can get back on this.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
